GetStream/stream-chat-swift
# Jump to unread messages - No interaction

### 🔗 Issue Links
https://github.com/GetStream/ios-issues-tracking/issues/345

### 🎯 Goal

Allow jumping to unread messages when the first unread message is not on screen.

### 🛠 Implementation

The whole logic around jump to unread and mark as read relies on `lastReadMessageId`. This is a value that is both sent and retrieved from backend. It determines the last message that the user has seen.

In order to adapt this to the necessities of the SDK, there is a bidirectional transformation from last read <-> first unread. Being the **first unread** the value that matters to the SDK. That way we can either scroll to the unread messages, or show a banner on top of that one to mark that the unread history starts from there.

Code context:
[getLastReadMessageId(firstUnreadMessageId: MessageId)](https://github.com/GetStream/stream-chat-swift/pull/2592/files/e6f6e181b9f7a55c131bbba29ca0f9072be4f15a#diff-6c459ba92353f66fbf686037e6921c360306e7238b4e83de2ac852c84b0c1c1aR1445)
[getFirstUnreadMessageId()](https://github.com/GetStream/stream-chat-swift/pull/2592/files/e6f6e181b9f7a55c131bbba29ca0f9072be4f15a#diff-6c459ba92353f66fbf686037e6921c360306e7238b4e83de2ac852c84b0c1c1aR1454)

### 🎨 Showcase

https://github.com/GetStream/stream-chat-swift/assets/7887319/4fa5eb49-a28b-4cce-8d46-478c43440ad0

### 🧪 Manual Testing Notes

1. Open a channel
2. Scroll to a message that is not on screen
3. Mark it as unread
4. Scroll to bottom

Expected result:
- A pill should be displayed on top showing the unread messages count.
- ~Tapping on the pill should jump to the first unread message~
- Tapping the X button should dismiss the pill

### ☑️ Contributor Checklist


### 🎁 Meme

![](https://media.giphy.com/media/Kv3bTmHLVLVzcW8eiv/giphy.gif)

### 🧱 Pending
- [ ] Update StreamChatUI Tests
